py3status/modules/window_manipulation.py

```diff
@@ -0,0 +1,396 @@
+                'border': w.border,
+                'is_focused': w.focused,
+                'is_urgent': w.urgent,
+                'is_floating': getattr(w, 'floating', False),  # python2
+                'is_scratchpad': is_scratchpad,
+            })
+
+        return new_scratchpad, new_window
+
+    def _manipulate_and_refresh(self, new_scratchpad=None, new_window=None):
+        """
+        For lack of a better word, we're trying to manipulate the data to make
+        the results more usable for the users. This is usually done by creating
+        new data and/or alter existing data. We strives to let data move through
+        without any or very little manipulation.
+        """
+        self.shared_data = new_window
+        count_floating = count_scratchpad = count_window = \
+            count_urgent = count_urgent_sp = 0
+
+        if new_scratchpad:
+            for s in new_scratchpad:
+                if s['is_urgent']:
+                    count_urgent_sp += 1
+
+        new_data = []
+        for w in new_window:
+            count_window += 1
+            # send all values (for focused_window) out to the format (as placeholders)
+            # so the users can do some simple things like printing name only.
+            if w['is_focused']:
+                self.id = w.get('id')
+                self.window_name = w.get('name')
+                self.class_name = w.get('class')
+            if w['is_scratchpad']:
+                count_scratchpad += 1
+            if w['is_urgent']:
+                count_urgent += 1
+
+            # when scrolling, we want to ignore this error because we already
+            # fixed it on update. we turned this into a boolean too.
+            try:
+                if 'on' in w['is_floating']:  # counts exclusively on updating
+                    w['is_floating'] = True
+                    count_floating += 1
+                else:
+                    w['is_floating'] = False
+            except:
+                if w['is_floating']:  # counts exclusively on scrolling
+                    count_floating += 1
+
+            new_data.append(self.py3.safe_format(self.format_window, w))
+
+        # send all counts out to the format (as placeholders)
+        self.count_floating = count_floating
+        self.count_scratchpad = count_scratchpad
+        self.count_urgent = count_urgent
+        self.count_urgent_sp = count_urgent_sp
+        self.count_window = count_window
+
+        # put everything together and do a refresh
+        format_separator = self.py3.safe_format(self.format_separator)
+        self.fmt_window = self.py3.composite_join(format_separator, new_data)
+        self.py3.update()
+
+    def _switch_window_manipulation(self, direction):
+        """
+        Users can scroll windows with mouse, click to focus. If an urgent occurs,
+        we will use a different click to focus. Disabled by default.
+        """
+        windows, f = self.shared_data, 'is_focused'
```

Comment:
`f` makes this function hard to read just use the literal.

also this is an evil double assignment thing which are just bad like this 

but you could consider simplifying things in this function eg
```diff
diff --git a/py3status/modules/window_manipulation.py b/py3status/modules/window_manipulation.py
index 0f4477a..0672952 100644
--- a/py3status/modules/window_manipulation.py
+++ b/py3status/modules/window_manipulation.py
@@ -203,6 +203,7 @@ class Py3status:
         t = Thread(target=self._loop)
         t.daemon = True
         t.start()
+        self.active = 0
 
     def _loop(self):
         def update_window_manipulation(i3, e=None):
@@ -287,10 +288,10 @@ class Py3status:
 
         new_data = []
         for w in new_window:
-            count_window += 1
             # send all values (for focused_window) out to the format (as placeholders)
             # so the users can do some simple things like printing name only.
             if w['is_focused']:
+                self.active = count_window
                 self.id = w.get('id')
                 self.window_name = w.get('name')
                 self.class_name = w.get('class')
@@ -299,6 +300,8 @@ class Py3status:
             if w['is_urgent']:
                 count_urgent += 1
 
+            count_window += 1
+
             # when scrolling, we want to ignore this error because we already
             # fixed it on update. we turned this into a boolean too.
             try:
@@ -330,22 +333,10 @@ class Py3status:
         Users can scroll windows with mouse, click to focus. If an urgent occurs,
         we will use a different click to focus. Disabled by default.
         """
-        windows, f = self.shared_data, 'is_focused'
-        length = len(windows)
-        for index, w in enumerate(windows):
-            if w.get(f):
-                windows[index][f] = False
-                if direction < 0:  # switch previous
-                    if index > 0:
-                        windows[index - 1][f] = True
-                    else:
-                        windows[index][f] = True
-                elif direction > 0:  # switch next
-                    if index < (length - 1):
-                        windows[index + 1][f] = True
-                    else:
-                        windows[length - 1][f] = True
-                break
+        windows = self.shared_data
+        self.active = (self.active + direction) % len(windows)
+        for index, window in enumerate(self.shared_data):
+            window['is_focused'] = (index == self.active)
 
         self._manipulate_and_refresh(new_window=windows)
 
```
slightly changes behaviour as we now loop round